core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory2.java

```diff
@@ -0,0 +1,256 @@
+      for (EquivalentAddressGroup addressGroup : removedAddrs) {
+        Subchannel subchannel = subchannels.remove(addressGroup);
+        subchannel.shutdown();
+      }
+
+      updatePicker(null);
+    }
+
+    @Override
+    public void handleNameResolutionError(Status error) {
+      updatePicker(error);
+    }
+
+    @Override
+    public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+      if (!subchannels.inverse().containsKey(subchannel)) {
+        return;
+      }
+      if (stateInfo.getState() == IDLE) {
+        subchannel.requestConnection();
+      }
+      setSubchannelStateInfo(subchannel, stateInfo);
+      updatePicker(getAggregatedError());
+    }
+
+    @Override
+    public void shutdown() {
+      for (Subchannel subchannel : getSubchannels()) {
+        subchannel.shutdown();
+      }
+    }
+
+    /**
+     * Updates picker with the list of active subchannels (state != TRANSIENT_ERROR).
+     */
+    private void updatePicker(@Nullable Status error) {
+      List<Subchannel> activeList = filterFailingSubchannels(getSubchannels());
+      helper.updatePicker(new Picker(activeList, error));
+    }
+
+    /**
+     * Filters out failing subchannels (state == TRANSIENT_ERROR).
+     */
+    private static List<Subchannel> filterFailingSubchannels(Collection<Subchannel> subchannels) {
+      List<Subchannel> nonFailingSubchannels = Lists.newArrayList();
+      for (Subchannel subchannel : subchannels) {
+        if (getSubchannelState(subchannel) != TRANSIENT_FAILURE) {
+          nonFailingSubchannels.add(subchannel);
+        }
+      }
+      return nonFailingSubchannels;
+    }
+
+    /**
+     * Converts list of {@link ResolvedServerInfoGroup} to {@link EquivalentAddressGroup} set.
+     */
+    private static Set<EquivalentAddressGroup> resolvedServerInfoGroupToEquivalentAddressGroup(
+        List<ResolvedServerInfoGroup> groupList) {
+      Set<EquivalentAddressGroup> addrs = new HashSet<EquivalentAddressGroup>(groupList.size());
+      for (ResolvedServerInfoGroup group : groupList) {
+        for (ResolvedServerInfo server : group.getResolvedServerInfoList()) {
+          addrs.add(server.toEquivalentAddressGroup());
+        }
+      }
+      return addrs;
+    }
+
+    /**
+     * If all subchannels are TRANSIENT_FAILURE, return the Status associated with an arbitrary
+     * subchannel otherwise, return null.
+     */
+    @Nullable
+    private Status getAggregatedError() {
+      for (Subchannel subchannel : subchannels.values()) {
+        if (getSubchannelState(subchannel) != TRANSIENT_FAILURE) {
+          return null;
+        }
+      }
+      return getSubchannelStatus(subchannels.values().iterator().next());
+    }
+
+    @VisibleForTesting
+    Collection<Subchannel> getSubchannels() {
+      return subchannels.values();
+    }
+
+    private static ConnectivityState getSubchannelState(Subchannel subchannel) {
```

Comment:
The added value of these three methods are basically the checkNotNull, which is repeated in them.  How about having one `AtomicReference<ConnectivityStateInfo> getSubchannelStateInfoRef(Subchannel subchannel)` instead of the three?